bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/grpc/async/AbstractRetryingOperation.java

```diff
@@ -165,20 +171,20 @@ protected void onError(Status status, Metadata trailers) {
 
     // Backoffs timed out.
     if (nextBackOff == BackOff.STOP) {
-      rpc.getRpcMetrics().markRetriesExhasted();
-      operationTimerContext.close();
-
-      String message = String.format("Exhausted retries after %d failures.", failedCount);
-      StatusRuntimeException cause = status.asRuntimeException();
-      setException(new BigtableRetriesExhaustedException(message, cause));
-      return;
+      setException(getExhaustedRetriesException(status));
     } else {
       String channelId = ChannelPool.extractIdentifier(trailers);
       LOG.info("Retrying failed call. Failure #%d, got: %s on channel %s",
           status.getCause(), failedCount, status, channelId);
+      performRetry(nextBackOff);
     }
+  }
 
-    performRetry(nextBackOff);
+  protected BigtableRetriesExhaustedException getExhaustedRetriesException(Status status) {
+    rpc.getRpcMetrics().markRetriesExhasted();
+    operationTimerContext.close();
```

Comment:
nit: Its a bit surprising to have these side effects in this method. I can't think of anything better